google/cloud/pubsub/internal/tracing_exactly_once_ack_handler_test.cc

```diff
@@ -0,0 +1,264 @@
+                         SpanWithStatus(opentelemetry::trace::StatusCode::kOk),
+                         SpanNamed("test-subscription nack"))));
+}
+
+TEST(TracingExactlyOnceAckHandlerTest, NackError) {
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_unique<MockExactlyOnceAckHandlerImpl>();
+  EXPECT_CALL(*mock, nack())
+      .WillOnce(Return(ByMove(
+          make_ready_future(Status{StatusCode::kPermissionDenied, "uh-oh"}))));
+  auto handler = MakeTestExactlyOnceAckHandler(std::move(mock));
+
+  EXPECT_THAT(std::move(handler)->nack().get(),
+              StatusIs(StatusCode::kPermissionDenied, "uh-oh"));
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(
+      spans,
+      Contains(AllOf(SpanHasInstrumentationScope(), SpanKindIsInternal(),
+                     SpanWithStatus(opentelemetry::trace::StatusCode::kError),
+                     SpanNamed("test-subscription nack"))));
+}
+
+TEST(TracingAckHandlerTest, NackAttributes) {
+  namespace sc = ::opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_unique<MockExactlyOnceAckHandlerImpl>();
+  EXPECT_CALL(*mock, nack())
+      .WillOnce(Return(ByMove(make_ready_future(Status{}))));
+  auto handler = MakeTestExactlyOnceAckHandler(std::move(mock));
+
+  EXPECT_THAT(std::move(handler)->nack().get(), StatusIs(StatusCode::kOk));
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(spans,
+              Contains(AllOf(SpanNamed("test-subscription nack"),
+                             SpanHasAttributes(OTelAttribute<std::string>(
+                                 sc::kMessagingSystem, "gcp_pubsub")))));
+  EXPECT_THAT(spans,
+              Contains(AllOf(SpanNamed("test-subscription nack"),
+                             SpanHasAttributes(OTelAttribute<std::string>(
+                                 "gcp.project_id", "test-project")))));
+  EXPECT_THAT(spans,
+              Contains(AllOf(SpanNamed("test-subscription nack"),
+                             SpanHasAttributes(OTelAttribute<std::string>(
+                                 sc::kMessagingOperation, "settle")))));
+  EXPECT_THAT(spans, Contains(AllOf(
+                         SpanNamed("test-subscription nack"),
+                         SpanHasAttributes(OTelAttribute<std::string>(
+                             sc::kCodeFunction, "pubsub::AckHandler::nack")))));
+  EXPECT_THAT(spans,
+              Contains(AllOf(
+                  SpanNamed("test-subscription nack"),
+                  SpanHasAttributes(OTelAttribute<std::int32_t>(
+                      "messaging.gcp_pubsub.message.delivery_attempt", 42)))));
+  EXPECT_THAT(spans,
+              Contains(AllOf(
+                  SpanNamed("test-subscription nack"),
+                  SpanHasAttributes(OTelAttribute<std::string>(
+                      sc::kMessagingDestinationName, "test-subscription")))));
```

Comment:
same comment about consolidating these attributes.

Reply:
I prefer having them separate so you know which line is failing. But it's fine either way. Done